src/compiler/checker.ts

```diff
@@ -25248,12 +25248,46 @@ namespace ts {
             }
         }
 
+        /**
+         * Try to find a resolved symbol for an expression without also resolving its type, as
+         * getSymbolAtLocation would (as that could be reentrant into contextual typing)
+         */
+         function getSymbolForExpression(e: Expression) {
+            if (e.symbol) {
+                return e.symbol;
+            }
+            if (isIdentifier(e)) {
+                return getResolvedSymbol(e);
+            }
+            if (isPropertyAccessExpression(e)) {
+                const lhsType = getTypeOfExpression(e.expression);
+                const prop = isPrivateIdentifier(e.name) ? tryGetPrivateIdentifierPropertyOfType(lhsType, e.name) : getPropertyOfType(lhsType, e.name.escapedText);
+                return prop;
+            }
+            return undefined;
+
+            function tryGetPrivateIdentifierPropertyOfType(type: Type, id: PrivateIdentifier) {
+                const lexicallyScopedSymbol = lookupSymbolForPrivateIdentifierDeclaration(id.escapedText, id);
+                const prop = lexicallyScopedSymbol && getPrivateIdentifierPropertyOfType(type, lexicallyScopedSymbol);
+                return prop;
+            }
+        }
+
         // In an assignment expression, the right operand is contextually typed by the type of the left operand.
         // Don't do this for assignment declarations unless there is a type tag on the assignment, to avoid circularity from checking the right operand.
         function getContextualTypeForAssignmentDeclaration(binaryExpression: BinaryExpression): Type | undefined {
             const kind = getAssignmentDeclarationKind(binaryExpression);
             switch (kind) {
                 case AssignmentDeclarationKind.None:

```

Comment:
I thought the problem was with `AssignmentDeclarationKind.ThisProperty`. Why is the new code in `None` instead? Is it because there's an existing PropertyDeclaration? 

The new code in `getSymbolForExpression` kind of looks like it's a fallback mode for when syntactic detection fails in the checker, which it shouldn't if it has succeeded in the binder. So maybe the fix needs to be to iron out some mismatch between the binder and checker.

Reply:
`ThisProperty` assignments don't exist in TS - or at least these aren't flagged as such. Moreover, when we don't find an annotation for those, we still fallback to the `getTypeOfExpression` call, which is what triggers a circularity warning. The symbol lookup code is essentially what `checkPropertyAccessExpression` does, except without eagerly materializing the type of the property. In the past, we've just done a simple `getPropertyOfType` inline, but in a post-private-symbol era, that's no longer sufficient.

Reply:
> `ThisProperty` assignments don't exist in TS - or at least these aren't flagged as such. 
Wellll, I guess that makes sense, since these technically evolve the type of an existing PropertyDeclaration. It's just a bit confusing since they're so similar.

However, I think they'll get marked as ThisProperty assignments in JS, so this PR needs a matching JS test to find out.